pkg/controller/controller.go

```diff
@@ -14,65 +14,92 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package server
+package controller
 
 import (
 	"errors"
 
 	"github.com/golang/glog"
 	"github.com/kubernetes-incubator/service-catalog/pkg/apis/servicecatalog"
-	"github.com/kubernetes-incubator/service-catalog/pkg/controller/catalog/util"
-	"github.com/kubernetes-incubator/service-catalog/pkg/controller/catalog/watch"
-
+	"github.com/kubernetes-incubator/service-catalog/pkg/controller/storage"
+	"github.com/kubernetes-incubator/service-catalog/pkg/controller/util"
+	"github.com/kubernetes-incubator/service-catalog/pkg/controller/watch"
 	k8swatch "k8s.io/client-go/1.5/pkg/watch"
 )
 
-type k8sHandler struct {
-	controller ServiceController
-	watcher    *watch.Watcher
+// Controller is an instance of the service catalog controller.
+type Controller struct {
+	controller *controller
 }
 
-func createK8sHandler(c ServiceController, w *watch.Watcher) (*k8sHandler, error) {
-	ret := &k8sHandler{
-		controller: c,
-		watcher:    w,
+// New creates an instance of the service catalog Controller.
+func New(w *watch.Watcher) (*Controller, error) {
+	h, err := createHandler(storage.CreateTPRStorage(w))
+	if err != nil {
+		glog.Errorf("Couldn't create controller: %v\n", err)
+		return nil, err
 	}
-	if w != nil {
-		glog.Infoln("Starting to watch for new Service Brokers")
-		err := w.Watch(watch.ServiceBroker, "default", ret.serviceBrokerCallback)
-		if err != nil {
-			glog.Errorf("Failed to start a watcher for Service Brokers: %v\n", err)
-		}
+	return &Controller{
+		controller: createController(h, w),
+	}, nil
+}
 
-		glog.Infoln("Starting to watch for new Service Instances")
-		err = w.Watch(watch.ServiceInstance, "default", ret.serviceInstanceCallback)
-		if err != nil {
-			glog.Errorf("Failed to start a watcher for Service Instances: %v\n", err)
-		}
+// Run starts the controller
+func (c *Controller) Run() {
+	c.controller.Run()
+}
 
-		glog.Infoln("Starting to watch for new Service Bindings")
-		err = w.Watch(watch.ServiceBinding, "default", ret.serviceBindingCallback)
-		if err != nil {
-			glog.Errorf("Failed to start a watcher for Service Bindings: %v\n", err)
-			return nil, err
-		}
-	} else {
+type controller struct {
+	handler Handler
+	watcher *watch.Watcher
+}
+
+func createController(h Handler, w *watch.Watcher) *controller {
+	return &controller{
+		handler: h,
+		watcher: w,
+	}
+}
+
+func (c *controller) Run() error {
+	if c.watcher == nil {
```

Comment:
Is this supposed to be concurrency-safe? I'm asking from curiosity, not blocking for this PR